### DISCUSSION TITLE
`@Value` 어노테이션 사용 시 주의할 점

### DISCUSSION THREAD

**Discussion by gyungchan-jo** [2024-woowacourse-study/level-interview#227] [Spring Core]
.

**Comment by J-I-H-O** [comment]
1. **주입 시점**
    
    `@Value` 어노테이션은 대상 컴포넌트가 스프링 빈으로 등록되고 의존 관계를 주입할 때 동작합니다. 
    따라서 환경 변수를 주입받는 대상 클래스에 `@Component` 어노테이션을 붙여주지 않는다면 해당 클래스는 컴포넌트 스캔이 대상이 되지 않아 스프링 빈으로 등록되지 않고, `@Value` 어노테이션 또한 동작하지 않습니다.
    
2. **상황에 따른 적절한 주입 방식 선택**
    
    빈을 주입받을 때와 마찬가지로, `@Value` 어노테이션을 사용할 때에도 필드 주입, 생성자 주입, setter 주입 등의 방식을 사용할 수 있습니다. 
    따라서 이들의 장단점을 비교하고, 상황에 따라 적절한 주입 방식을 선택해야 합니다.
    

1. **프로퍼티 파일의 경로와 스코프**
    
    프로퍼티 파일의 경로가 정확한지 또한 확인해야 합니다. `application.yaml` 이 클래스 패스에 존재해야 하고, 프로퍼티 파일이 여러 개일 경우 우선순위를 고려해야 합니다.

**Comment by le2sky** [comment]
`@Value` 어노테이션을 이용하면 스프링 프레임워크에게 프로퍼티 값을 주입받을 수 있습니다.
해당 어노테이션을 이용할때는 다음과 같은 점을 주의해야합니다.

- 프로퍼티 키를 문자열로 작성하기 때문에 오타로 인해 프로퍼티를 찾지 못하는 문제가 생길 수 있습니다.
- 주입 받는 클래스는 `@Component`를 추가하여 빈으로 등록되어 있어야한다는 것을 주의해야합니다.(`@Configuration`도 가능)
- 프로퍼티에 명시된 값의 타입과 주입받을 필드나 파라미터 타입이 다른 경우 변환이 정상적으로 이루어지지 않습니다.
